Library/MeterGeometry.cpp

```diff
@@ -0,0 +1,370 @@
+	m_Shapes(),
+	m_MeasureOptions(),
+	m_NeedsRedraw(false)
+{
+}
+
+MeterGeometry::~MeterGeometry()
+{
+}
+
+void MeterGeometry::Initialize()
+{
+	Meter::Initialize();
+}
+
+bool MeterGeometry::ParseShape(GeometryShape & shape, const LPCWSTR & optionName, const LPCWSTR & optionValue)
+{
+	if (_wcsicmp(optionName, L"Rectangle") == 0) shape.m_Shape = ParseRect(shape, ConfigParser::ParseRECT(optionValue));
+	else return false;
+	//Easy to do with a rectangle, but this will calculate the bounds of all shape types
+	if (shape.m_Shape) {
+		shape.m_Shape->GetBounds(D2D1::Matrix3x2F::Identity(), &shape.m_Bounds);
+		if (!m_WDefined && m_W < shape.m_Bounds.right + std::ceilf(shape.m_OutlineWidth / 4 * 3) + GetWidthPadding()) {
+			m_W = std::ceil(shape.m_Bounds.right + shape.m_OutlineWidth / 4 * 3 + GetWidthPadding());
+			if (!IsHidden())
+				m_Skin->SetResizeWindowMode(RESIZEMODE_CHECK);
+		}
+		if (!m_HDefined && m_H < shape.m_Bounds.bottom + std::ceilf(shape.m_OutlineWidth / 4 * 3) + GetHeightPadding()) {
+			m_H = std::ceil(shape.m_Bounds.bottom + shape.m_OutlineWidth / 4 * 3 + GetHeightPadding());
+			if(!IsHidden())
+				m_Skin->SetResizeWindowMode(RESIZEMODE_CHECK);
+		}
+	}
+	return true;
+}
+
+bool MeterGeometry::ReplaceShapeOption(GeometryShape & shape, const LPCWSTR & optionName, const LPCWSTR & optionValue)
+{
+	if (_wcsicmp(optionName, L"FillColor") == 0)			shape.m_FillColor = ToColorF(ConfigParser::ParseColor(optionValue));
+	else if (_wcsicmp(optionName, L"OutlineWidth") == 0)	shape.m_OutlineWidth = (float)ConfigParser::ParseDouble(optionValue, 1.0f);
+	else if (_wcsicmp(optionName, L"OutlineColor") == 0)	shape.m_OutlineColor = ToColorF(ConfigParser::ParseColor(optionValue));
+	else if (_wcsicmp(optionName, L"Offset") == 0)			shape.m_Offset = ParseSize(optionName, 0);
+	else if (_wcsicmp(optionName, L"Scale") == 0)			shape.m_Scale = ParseSize(optionName, 1);
+	else if (_wcsicmp(optionName, L"Skew") == 0)			shape.m_Skew = ParsePoint(optionName, 0);
+	else if (_wcsicmp(optionName, L"Rotation") == 0)		shape.m_Rotation = ParseRotation(optionName, 0, shape);
+	else return false;
+	return true;
+}
+
+bool MeterGeometry::IsPostOption(const LPCWSTR & option)
+{
+	return _wcsicmp(option, L"Rotation") == 0;
+}
+
+bool MeterGeometry::IsShape(const LPCWSTR & option)
+{
+	return _wcsicmp(option, L"Rectangle") == 0;
+}
+void MeterGeometry::ReadOptions(ConfigParser & parser, const WCHAR * section)
+{
+	Meter::ReadOptions(parser, section);
+	//Setting Width and Height to 0 to stop padding from consuming everything
+	if (!m_WDefined)
+		m_W = 0;
+	if (!m_HDefined)
+		m_H = 0;
+
+	m_Shapes.clear();
+	m_MeasureOptions.clear();
+	size_t currentShapeId = 1;
+	std::wstring shapeName = L"Shape";
+	std::wstring shapeOption = parser.ReadString(section, L"Shape", L"");
+	std::map<LPCWSTR, std::pair<LPCWSTR, LPCWSTR>> postOptions;
+	if (!shapeOption.empty()) 
+		do
+		{
+			std::vector<std::wstring> shapeTokens = CustomTokenize(shapeOption.c_str(), L"|:");
```

Comment:
I'm not sure this is a good approach because it means you can mix and match `:` and `|`. Maybe you could first split by `|` and then split each of those tokens by `:`?